// @vechain/connex-framework/esm/filter.js
import * as R from './rules';
const MAX_LIMIT = 256;
export function newFilter(readyDriver, kind, criteria) {
    const filterBody = {
        range: {
            unit: 'block',
            from: 0,
            to: Math.pow(2, 32) - 1
        },
        options: {
            offset: 0,
            limit: 10
        },
        criteriaSet: criteria,
        order: 'asc'
    };
    let cacheHints;
    return {
        range(range) {
            R.test(range, {
                unit: v => (v === 'block' || v === 'time') ? '' : `expected 'block' or 'time'`,
                from: R.uint64,
                to: R.uint64
            }, 'arg0');
            R.ensure(range.from <= range.to, 'arg0.from: expected <= arg0.to');
            filterBody.range = Object.assign({}, range);
            return this;
        },
        order(order) {
            R.ensure(order === 'asc' || order === 'desc', `arg0: expected 'asc' or 'desc'`);
            filterBody.order = order;
            return this;
        },
        cache(hints) {
            cacheHints = R.test(hints, [R.address], 'arg0').map(t => t.toLowerCase());
            return this;
        },
        apply(offset, limit) {
            R.test(offset, R.uint64, 'arg0');
            R.ensure(limit >= 0 && limit <= MAX_LIMIT && Number.isInteger(limit), `arg1: expected unsigned integer <= ${MAX_LIMIT}`);
            filterBody.options.offset = offset;
            filterBody.options.limit = limit;
            if (kind === 'transfer') {
                return readyDriver.then(d => d.filterTransferLogs(filterBody, cacheHints));
            }
            else {
                return readyDriver.then(d => d.filterEventLogs(filterBody, cacheHints));
            }
        }
    };
}
