// @vechain/connex-framework/esm/rules.js
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
export function isDecString(val) {
    return typeof val === 'string' && /^[0-9]+$/.test(val);
}
export function isHexString(val) {
    return typeof val === 'string' && /^0x[0-9a-f]+$/i.test(val);
}
export function isHexBytes(val, n) {
    if (typeof val !== 'string' || !/^0x[0-9a-f]*$/i.test(val)) {
        return false;
    }
    return n ? val.length === n * 2 + 2 : val.length % 2 === 0;
}
export function isUInt(val, bit) {
    if (val < 0 || !Number.isInteger(val)) {
        return false;
    }
    return bit ? val < Math.pow(2, bit) : true;
}
export function isBigInt(v) {
    return typeof v === 'string' ?
        (isDecString(v) || isHexString(v)) :
        isUInt(v, 0);
}
export class BadParameter extends Error {
    constructor(msg) {
        super(msg);
    }
}
BadParameter.prototype.name = 'BadParameter';
export function ensure(b, msg) {
    if (!b) {
        throw new BadParameter(msg);
    }
}
import * as V from 'validator-ts';
export function test(value, scheme, context) {
    try {
        return V.validate(value, scheme, context);
    }
    catch (err) {
        if (err instanceof V.ValidationError) {
            throw new BadParameter(err.message);
        }
        throw err;
    }
}
// rules
export function bytes(v) {
    return isHexBytes(v) ? '' : 'expected bytes in hex string';
}
export function bytes8(v) {
    return isHexBytes(v, 8) ? '' : 'expected bytes8';
}
export function bytes32(v) {
    return isHexBytes(v, 32) ? '' : 'expected bytes32';
}
export function uint8(v) {
    return isUInt(v, 8) ? '' : 'expected 8-bit unsigned integer';
}
export function uint32(v) {
    return isUInt(v, 32) ? '' : 'expected 32-bit unsigned integer';
}
export function uint64(v) {
    return isUInt(v, 64) ? '' : 'expected 64-bit unsigned integer';
}
export function bool(v) {
    return typeof v === 'boolean' ? '' : 'expected boolean';
}
export function bigInt(v) {
    return isBigInt(v) ? '' : 'expected unsigned integer in number or string';
}
export function hexString(v) {
    return isHexString(v) ? '' : 'expected integer in hex string';
}
export function address(v) {
    return isHexBytes(v, 20) ? '' : 'expected address';
}
export function string(v) {
    return typeof v === 'string' ? '' : 'expected string';
}
