// @vechain/connex-framework/esm/vendor.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as R from './rules';
import * as V from 'validator-ts';
import { abi } from 'thor-devkit';
import BigNumber from 'bignumber.js';
export function newVendor(driver) {
    return {
        sign: (kind, msg) => {
            if (kind === 'tx') {
                R.test(msg, [clauseScheme], 'arg1');
                return newTxSigningService(Promise.resolve(driver), msg);
            }
            else if (kind === 'cert') {
                R.test(msg, {
                    purpose: v => (v === 'agreement' || v === 'identification') ?
                        '' : `expected 'agreement' or 'identification'`,
                    payload: {
                        type: v => v === 'text' ? '' : `expected 'text'`,
                        content: R.string
                    }
                }, 'arg1');
                return newCertSigningService(Promise.resolve(driver), msg);
            }
            else {
                throw new R.BadParameter(`arg0: expected 'tx' or 'cert'`);
            }
        }
    };
}
export function newTxSigningService(readyDriver, msg) {
    const opts = {};
    return {
        signer(addr) {
            opts.signer = R.test(addr, R.address, 'arg0').toLowerCase();
            return this;
        },
        gas(gas) {
            opts.gas = R.test(gas, R.uint64, 'arg0');
            return this;
        },
        dependsOn(txid) {
            opts.dependsOn = R.test(txid, R.bytes32, 'arg0').toLowerCase();
            return this;
        },
        link(url) {
            opts.link = R.test(url, R.string, 'arg0');
            return this;
        },
        comment(text) {
            opts.comment = R.test(text, R.string, 'arg0');
            return this;
        },
        delegate(url, signer) {
            R.ensure(typeof url === 'string', `arg0: expected url string`);
            R.test(signer, V.optional(R.address), 'arg1');
            opts.delegator = { url, signer: signer && signer.toLowerCase() };
            return this;
        },
        accepted(cb) {
            R.ensure(typeof cb === 'function', 'arg0: expected function');
            opts.onAccepted = cb;
            return this;
        },
        request() {
            const transformedMsg = msg.map(c => (Object.assign(Object.assign({}, c), { value: new BigNumber(c.value).toString(10) })));
            return (() => __awaiter(this, void 0, void 0, function* () {
                try {
                    const driver = yield readyDriver;
                    return yield driver.signTx(transformedMsg, opts);
                }
                catch (err) {
                    throw new Rejected(err.message);
                }
            }))();
        }
    };
}
function newCertSigningService(readyDriver, msg) {
    const opts = {};
    return {
        signer(addr) {
            opts.signer = R.test(addr, R.address, 'arg0').toLowerCase();
            return this;
        },
        link(url) {
            opts.link = R.test(url, R.string, 'arg0');
            return this;
        },
        accepted(cb) {
            R.ensure(typeof cb === 'function', 'arg0: expected function');
            opts.onAccepted = cb;
            return this;
        },
        request() {
            return (() => __awaiter(this, void 0, void 0, function* () {
                try {
                    const driver = yield readyDriver;
                    return yield driver.signCert(msg, opts);
                }
                catch (err) {
                    throw new Rejected(err.message);
                }
            }))();
        }
    };
}
class Rejected extends Error {
    constructor(msg) {
        super(msg);
    }
}
Rejected.prototype.name = 'Rejected';
const clauseScheme = {
    to: V.nullable(R.address),
    value: R.bigInt,
    data: V.optional(R.bytes),
    comment: V.optional(R.string),
    abi: V.optional(v => {
        if (!(v instanceof Object)) {
            return 'expected object';
        }
        try {
            new abi.Function(v).signature;
            return '';
        }
        catch (err) {
            return `expected valid ABI (${err.message})`;
        }
    })
};
