// secretjs/dist/protobuf_stuff/tendermint/state/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.State = exports.Version = exports.ConsensusParamsInfo = exports.ValidatorsInfo = exports.ABCIResponses = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const _m0 = __importStar(require("protobufjs/minimal"));
const types_1 = require("../../tendermint/abci/types");
const validator_1 = require("../../tendermint/types/validator");
const params_1 = require("../../tendermint/types/params");
const types_2 = require("../../tendermint/version/types");
const types_3 = require("../../tendermint/types/types");
const timestamp_1 = require("../../google/protobuf/timestamp");
exports.protobufPackage = "tendermint.state";
function createBaseABCIResponses() {
    return { deliverTxs: [], endBlock: undefined, beginBlock: undefined };
}
exports.ABCIResponses = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.deliverTxs) {
            types_1.ResponseDeliverTx.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.endBlock !== undefined) {
            types_1.ResponseEndBlock.encode(message.endBlock, writer.uint32(18).fork()).ldelim();
        }
        if (message.beginBlock !== undefined) {
            types_1.ResponseBeginBlock.encode(message.beginBlock, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseABCIResponses();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.deliverTxs.push(types_1.ResponseDeliverTx.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.endBlock = types_1.ResponseEndBlock.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.beginBlock = types_1.ResponseBeginBlock.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            deliverTxs: Array.isArray(object === null || object === void 0 ? void 0 : object.deliverTxs)
                ? object.deliverTxs.map((e) => types_1.ResponseDeliverTx.fromJSON(e))
                : [],
            endBlock: isSet(object.endBlock)
                ? types_1.ResponseEndBlock.fromJSON(object.endBlock)
                : undefined,
            beginBlock: isSet(object.beginBlock)
                ? types_1.ResponseBeginBlock.fromJSON(object.beginBlock)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.deliverTxs) {
            obj.deliverTxs = message.deliverTxs.map((e) => e ? types_1.ResponseDeliverTx.toJSON(e) : undefined);
        }
        else {
            obj.deliverTxs = [];
        }
        message.endBlock !== undefined &&
            (obj.endBlock = message.endBlock
                ? types_1.ResponseEndBlock.toJSON(message.endBlock)
                : undefined);
        message.beginBlock !== undefined &&
            (obj.beginBlock = message.beginBlock
                ? types_1.ResponseBeginBlock.toJSON(message.beginBlock)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseABCIResponses();
        message.deliverTxs =
            ((_a = object.deliverTxs) === null || _a === void 0 ? void 0 : _a.map((e) => types_1.ResponseDeliverTx.fromPartial(e))) || [];
        message.endBlock =
            object.endBlock !== undefined && object.endBlock !== null
                ? types_1.ResponseEndBlock.fromPartial(object.endBlock)
                : undefined;
        message.beginBlock =
            object.beginBlock !== undefined && object.beginBlock !== null
                ? types_1.ResponseBeginBlock.fromPartial(object.beginBlock)
                : undefined;
        return message;
    },
};
function createBaseValidatorsInfo() {
    return { validatorSet: undefined, lastHeightChanged: "0" };
}
exports.ValidatorsInfo = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.validatorSet !== undefined) {
            validator_1.ValidatorSet.encode(message.validatorSet, writer.uint32(10).fork()).ldelim();
        }
        if (message.lastHeightChanged !== "0") {
            writer.uint32(16).int64(message.lastHeightChanged);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseValidatorsInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.validatorSet = validator_1.ValidatorSet.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.lastHeightChanged = longToString(reader.int64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            validatorSet: isSet(object.validatorSet)
                ? validator_1.ValidatorSet.fromJSON(object.validatorSet)
                : undefined,
            lastHeightChanged: isSet(object.lastHeightChanged)
                ? String(object.lastHeightChanged)
                : "0",
        };
    },
    toJSON(message) {
        const obj = {};
        message.validatorSet !== undefined &&
            (obj.validatorSet = message.validatorSet
                ? validator_1.ValidatorSet.toJSON(message.validatorSet)
                : undefined);
        message.lastHeightChanged !== undefined &&
            (obj.lastHeightChanged = message.lastHeightChanged);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseValidatorsInfo();
        message.validatorSet =
            object.validatorSet !== undefined && object.validatorSet !== null
                ? validator_1.ValidatorSet.fromPartial(object.validatorSet)
                : undefined;
        message.lastHeightChanged = (_a = object.lastHeightChanged) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
function createBaseConsensusParamsInfo() {
    return { consensusParams: undefined, lastHeightChanged: "0" };
}
exports.ConsensusParamsInfo = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.consensusParams !== undefined) {
            params_1.ConsensusParams.encode(message.consensusParams, writer.uint32(10).fork()).ldelim();
        }
        if (message.lastHeightChanged !== "0") {
            writer.uint32(16).int64(message.lastHeightChanged);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConsensusParamsInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.consensusParams = params_1.ConsensusParams.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.lastHeightChanged = longToString(reader.int64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            consensusParams: isSet(object.consensusParams)
                ? params_1.ConsensusParams.fromJSON(object.consensusParams)
                : undefined,
            lastHeightChanged: isSet(object.lastHeightChanged)
                ? String(object.lastHeightChanged)
                : "0",
        };
    },
    toJSON(message) {
        const obj = {};
        message.consensusParams !== undefined &&
            (obj.consensusParams = message.consensusParams
                ? params_1.ConsensusParams.toJSON(message.consensusParams)
                : undefined);
        message.lastHeightChanged !== undefined &&
            (obj.lastHeightChanged = message.lastHeightChanged);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseConsensusParamsInfo();
        message.consensusParams =
            object.consensusParams !== undefined && object.consensusParams !== null
                ? params_1.ConsensusParams.fromPartial(object.consensusParams)
                : undefined;
        message.lastHeightChanged = (_a = object.lastHeightChanged) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
function createBaseVersion() {
    return { consensus: undefined, software: "" };
}
exports.Version = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.consensus !== undefined) {
            types_2.Consensus.encode(message.consensus, writer.uint32(10).fork()).ldelim();
        }
        if (message.software !== "") {
            writer.uint32(18).string(message.software);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseVersion();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.consensus = types_2.Consensus.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.software = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            consensus: isSet(object.consensus)
                ? types_2.Consensus.fromJSON(object.consensus)
                : undefined,
            software: isSet(object.software) ? String(object.software) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.consensus !== undefined &&
            (obj.consensus = message.consensus
                ? types_2.Consensus.toJSON(message.consensus)
                : undefined);
        message.software !== undefined && (obj.software = message.software);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseVersion();
        message.consensus =
            object.consensus !== undefined && object.consensus !== null
                ? types_2.Consensus.fromPartial(object.consensus)
                : undefined;
        message.software = (_a = object.software) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseState() {
    return {
        version: undefined,
        chainId: "",
        initialHeight: "0",
        lastBlockHeight: "0",
        lastBlockId: undefined,
        lastBlockTime: undefined,
        nextValidators: undefined,
        validators: undefined,
        lastValidators: undefined,
        lastHeightValidatorsChanged: "0",
        consensusParams: undefined,
        lastHeightConsensusParamsChanged: "0",
        lastResultsHash: new Uint8Array(),
        appHash: new Uint8Array(),
    };
}
exports.State = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.version !== undefined) {
            exports.Version.encode(message.version, writer.uint32(10).fork()).ldelim();
        }
        if (message.chainId !== "") {
            writer.uint32(18).string(message.chainId);
        }
        if (message.initialHeight !== "0") {
            writer.uint32(112).int64(message.initialHeight);
        }
        if (message.lastBlockHeight !== "0") {
            writer.uint32(24).int64(message.lastBlockHeight);
        }
        if (message.lastBlockId !== undefined) {
            types_3.BlockID.encode(message.lastBlockId, writer.uint32(34).fork()).ldelim();
        }
        if (message.lastBlockTime !== undefined) {
            timestamp_1.Timestamp.encode(message.lastBlockTime, writer.uint32(42).fork()).ldelim();
        }
        if (message.nextValidators !== undefined) {
            validator_1.ValidatorSet.encode(message.nextValidators, writer.uint32(50).fork()).ldelim();
        }
        if (message.validators !== undefined) {
            validator_1.ValidatorSet.encode(message.validators, writer.uint32(58).fork()).ldelim();
        }
        if (message.lastValidators !== undefined) {
            validator_1.ValidatorSet.encode(message.lastValidators, writer.uint32(66).fork()).ldelim();
        }
        if (message.lastHeightValidatorsChanged !== "0") {
            writer.uint32(72).int64(message.lastHeightValidatorsChanged);
        }
        if (message.consensusParams !== undefined) {
            params_1.ConsensusParams.encode(message.consensusParams, writer.uint32(82).fork()).ldelim();
        }
        if (message.lastHeightConsensusParamsChanged !== "0") {
            writer.uint32(88).int64(message.lastHeightConsensusParamsChanged);
        }
        if (message.lastResultsHash.length !== 0) {
            writer.uint32(98).bytes(message.lastResultsHash);
        }
        if (message.appHash.length !== 0) {
            writer.uint32(106).bytes(message.appHash);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.version = exports.Version.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.chainId = reader.string();
                    break;
                case 14:
                    message.initialHeight = longToString(reader.int64());
                    break;
                case 3:
                    message.lastBlockHeight = longToString(reader.int64());
                    break;
                case 4:
                    message.lastBlockId = types_3.BlockID.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.lastBlockTime = timestamp_1.Timestamp.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.nextValidators = validator_1.ValidatorSet.decode(reader, reader.uint32());
                    break;
                case 7:
                    message.validators = validator_1.ValidatorSet.decode(reader, reader.uint32());
                    break;
                case 8:
                    message.lastValidators = validator_1.ValidatorSet.decode(reader, reader.uint32());
                    break;
                case 9:
                    message.lastHeightValidatorsChanged = longToString(reader.int64());
                    break;
                case 10:
                    message.consensusParams = params_1.ConsensusParams.decode(reader, reader.uint32());
                    break;
                case 11:
                    message.lastHeightConsensusParamsChanged = longToString(reader.int64());
                    break;
                case 12:
                    message.lastResultsHash = reader.bytes();
                    break;
                case 13:
                    message.appHash = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            version: isSet(object.version)
                ? exports.Version.fromJSON(object.version)
                : undefined,
            chainId: isSet(object.chainId) ? String(object.chainId) : "",
            initialHeight: isSet(object.initialHeight)
                ? String(object.initialHeight)
                : "0",
            lastBlockHeight: isSet(object.lastBlockHeight)
                ? String(object.lastBlockHeight)
                : "0",
            lastBlockId: isSet(object.lastBlockId)
                ? types_3.BlockID.fromJSON(object.lastBlockId)
                : undefined,
            lastBlockTime: isSet(object.lastBlockTime)
                ? fromJsonTimestamp(object.lastBlockTime)
                : undefined,
            nextValidators: isSet(object.nextValidators)
                ? validator_1.ValidatorSet.fromJSON(object.nextValidators)
                : undefined,
            validators: isSet(object.validators)
                ? validator_1.ValidatorSet.fromJSON(object.validators)
                : undefined,
            lastValidators: isSet(object.lastValidators)
                ? validator_1.ValidatorSet.fromJSON(object.lastValidators)
                : undefined,
            lastHeightValidatorsChanged: isSet(object.lastHeightValidatorsChanged)
                ? String(object.lastHeightValidatorsChanged)
                : "0",
            consensusParams: isSet(object.consensusParams)
                ? params_1.ConsensusParams.fromJSON(object.consensusParams)
                : undefined,
            lastHeightConsensusParamsChanged: isSet(object.lastHeightConsensusParamsChanged)
                ? String(object.lastHeightConsensusParamsChanged)
                : "0",
            lastResultsHash: isSet(object.lastResultsHash)
                ? bytesFromBase64(object.lastResultsHash)
                : new Uint8Array(),
            appHash: isSet(object.appHash)
                ? bytesFromBase64(object.appHash)
                : new Uint8Array(),
        };
    },
    toJSON(message) {
        const obj = {};
        message.version !== undefined &&
            (obj.version = message.version
                ? exports.Version.toJSON(message.version)
                : undefined);
        message.chainId !== undefined && (obj.chainId = message.chainId);
        message.initialHeight !== undefined &&
            (obj.initialHeight = message.initialHeight);
        message.lastBlockHeight !== undefined &&
            (obj.lastBlockHeight = message.lastBlockHeight);
        message.lastBlockId !== undefined &&
            (obj.lastBlockId = message.lastBlockId
                ? types_3.BlockID.toJSON(message.lastBlockId)
                : undefined);
        message.lastBlockTime !== undefined &&
            (obj.lastBlockTime = fromTimestamp(message.lastBlockTime).toISOString());
        message.nextValidators !== undefined &&
            (obj.nextValidators = message.nextValidators
                ? validator_1.ValidatorSet.toJSON(message.nextValidators)
                : undefined);
        message.validators !== undefined &&
            (obj.validators = message.validators
                ? validator_1.ValidatorSet.toJSON(message.validators)
                : undefined);
        message.lastValidators !== undefined &&
            (obj.lastValidators = message.lastValidators
                ? validator_1.ValidatorSet.toJSON(message.lastValidators)
                : undefined);
        message.lastHeightValidatorsChanged !== undefined &&
            (obj.lastHeightValidatorsChanged = message.lastHeightValidatorsChanged);
        message.consensusParams !== undefined &&
            (obj.consensusParams = message.consensusParams
                ? params_1.ConsensusParams.toJSON(message.consensusParams)
                : undefined);
        message.lastHeightConsensusParamsChanged !== undefined &&
            (obj.lastHeightConsensusParamsChanged =
                message.lastHeightConsensusParamsChanged);
        message.lastResultsHash !== undefined &&
            (obj.lastResultsHash = base64FromBytes(message.lastResultsHash !== undefined
                ? message.lastResultsHash
                : new Uint8Array()));
        message.appHash !== undefined &&
            (obj.appHash = base64FromBytes(message.appHash !== undefined ? message.appHash : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBaseState();
        message.version =
            object.version !== undefined && object.version !== null
                ? exports.Version.fromPartial(object.version)
                : undefined;
        message.chainId = (_a = object.chainId) !== null && _a !== void 0 ? _a : "";
        message.initialHeight = (_b = object.initialHeight) !== null && _b !== void 0 ? _b : "0";
        message.lastBlockHeight = (_c = object.lastBlockHeight) !== null && _c !== void 0 ? _c : "0";
        message.lastBlockId =
            object.lastBlockId !== undefined && object.lastBlockId !== null
                ? types_3.BlockID.fromPartial(object.lastBlockId)
                : undefined;
        message.lastBlockTime =
            object.lastBlockTime !== undefined && object.lastBlockTime !== null
                ? timestamp_1.Timestamp.fromPartial(object.lastBlockTime)
                : undefined;
        message.nextValidators =
            object.nextValidators !== undefined && object.nextValidators !== null
                ? validator_1.ValidatorSet.fromPartial(object.nextValidators)
                : undefined;
        message.validators =
            object.validators !== undefined && object.validators !== null
                ? validator_1.ValidatorSet.fromPartial(object.validators)
                : undefined;
        message.lastValidators =
            object.lastValidators !== undefined && object.lastValidators !== null
                ? validator_1.ValidatorSet.fromPartial(object.lastValidators)
                : undefined;
        message.lastHeightValidatorsChanged =
            (_d = object.lastHeightValidatorsChanged) !== null && _d !== void 0 ? _d : "0";
        message.consensusParams =
            object.consensusParams !== undefined && object.consensusParams !== null
                ? params_1.ConsensusParams.fromPartial(object.consensusParams)
                : undefined;
        message.lastHeightConsensusParamsChanged =
            (_e = object.lastHeightConsensusParamsChanged) !== null && _e !== void 0 ? _e : "0";
        message.lastResultsHash = (_f = object.lastResultsHash) !== null && _f !== void 0 ? _f : new Uint8Array();
        message.appHash = (_g = object.appHash) !== null && _g !== void 0 ? _g : new Uint8Array();
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
const atob = globalThis.atob ||
    ((b64) => globalThis.Buffer.from(b64, "base64").toString("binary"));
function bytesFromBase64(b64) {
    const bin = atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
const btoa = globalThis.btoa ||
    ((bin) => globalThis.Buffer.from(bin, "binary").toString("base64"));
function base64FromBytes(arr) {
    const bin = [];
    for (const byte of arr) {
        bin.push(String.fromCharCode(byte));
    }
    return btoa(bin.join(""));
}
function toTimestamp(date) {
    const seconds = Math.trunc(date.getTime() / 1000).toString();
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = Number(t.seconds) * 1000;
    millis += t.nanos / 1000000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return toTimestamp(o);
    }
    else if (typeof o === "string") {
        return toTimestamp(new Date(o));
    }
    else {
        return timestamp_1.Timestamp.fromJSON(o);
    }
}
function longToString(long) {
    return long.toString();
}
if (_m0.util.Long !== long_1.default) {
    _m0.util.Long = long_1.default;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
