// crypto-exchange-rate/dist/model/domain/SupportedCurrency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedCurrency = void 0;
var SupportedCurrency;
(function (SupportedCurrency) {
    SupportedCurrency["XPNET"] = "XPNET";
    SupportedCurrency["EGLD"] = "EGLD";
    SupportedCurrency["HT"] = "HT";
    SupportedCurrency["BNB"] = "BNB";
    SupportedCurrency["ETH"] = "ETH";
    SupportedCurrency["AVAX"] = "AVAX";
    SupportedCurrency["MATIC"] = "MATIC";
    SupportedCurrency["FTM"] = "FTM";
    SupportedCurrency["ONE"] = "ONE";
    SupportedCurrency["TRX"] = "TRX";
    SupportedCurrency["CELO"] = "CELO";
    SupportedCurrency["ONT"] = "ONT";
    SupportedCurrency["XDAI"] = "XDAI";
    SupportedCurrency["ALGO"] = "ALGO";
    SupportedCurrency["FUSE"] = "FUSE";
    SupportedCurrency["OPL"] = "OPL";
    SupportedCurrency["XTZ"] = "XTZ";
    SupportedCurrency["VLX"] = "VLX";
    SupportedCurrency["IOTX"] = "IOTX";
    SupportedCurrency["AURORA"] = "ETH";
    SupportedCurrency["VET"] = "VET";
    SupportedCurrency["CKB"] = "CKB";
    SupportedCurrency["GT"] = "GT";
    SupportedCurrency["SCRT"] = "SCRT";
})(SupportedCurrency = exports.SupportedCurrency || (exports.SupportedCurrency = {}));
