// validator-ts/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * convert the given Scheme into a Rule that allows undefined value
 * @param scheme the scheme
 */
function optional(scheme) {
    return (value, context) => {
        if (value !== undefined) {
            validate(value, scheme, context);
        }
        return '';
    };
}
exports.optional = optional;
/**
 * convert the given Scheme into a Rule that allows null value
 * @param scheme the scheme
 */
function nullable(scheme) {
    return (value, context) => {
        if (value !== null) {
            validate(value, scheme, context);
        }
        return '';
    };
}
exports.nullable = nullable;
/** Validator class */
class Validator {
    constructor(scheme) {
        this.scheme = scheme;
    }
    test(value, context) {
        return validate(value, this.scheme, context);
    }
}
exports.Validator = Validator;
/** Error class describes validation error */
class ValidationError extends Error {
    constructor(rawMessage, context) {
        super(context ? `${context}: ${rawMessage}` : rawMessage);
        this.rawMessage = rawMessage;
        this.context = context;
    }
}
exports.ValidationError = ValidationError;
ValidationError.prototype.name = 'ValidationError';
/**
 * direct function to validate value without construct Validator object
 * @param value value to be validated
 * @param scheme scheme
 * @param context context string appears in error object
 */
function validate(value, scheme, context) {
    if (Array.isArray(scheme)) {
        if (!Array.isArray(value)) {
            throw new ValidationError('expected array', context || '');
        }
        value.forEach((el, i) => validate(el, scheme[0], context ? `${context}.#${i}` : `#${i}`));
    }
    else if (scheme instanceof Function) {
        const errMsg = scheme(value, context);
        if (errMsg) {
            throw new ValidationError(errMsg, context || '');
        }
    }
    else {
        if (!(value instanceof Object)) {
            throw new ValidationError('expected object', context || '');
        }
        for (const key in scheme) {
            if (scheme.hasOwnProperty(key)) {
                validate(value[key], scheme[key], context ? `${context}.${key}` : key);
            }
        }
    }
    return value;
}
exports.validate = validate;
