// @temple-wallet/dapp/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplePageMessageType = exports.TempleDAppErrorType = exports.TempleDAppMessageType = void 0;
var TempleDAppMessageType;
(function (TempleDAppMessageType) {
    TempleDAppMessageType["GetCurrentPermissionRequest"] = "GET_CURRENT_PERMISSION_REQUEST";
    TempleDAppMessageType["GetCurrentPermissionResponse"] = "GET_CURRENT_PERMISSION_RESPONSE";
    TempleDAppMessageType["PermissionRequest"] = "PERMISSION_REQUEST";
    TempleDAppMessageType["PermissionResponse"] = "PERMISSION_RESPONSE";
    TempleDAppMessageType["OperationRequest"] = "OPERATION_REQUEST";
    TempleDAppMessageType["OperationResponse"] = "OPERATION_RESPONSE";
    TempleDAppMessageType["SignRequest"] = "SIGN_REQUEST";
    TempleDAppMessageType["SignResponse"] = "SIGN_RESPONSE";
    TempleDAppMessageType["BroadcastRequest"] = "BROADCAST_REQUEST";
    TempleDAppMessageType["BroadcastResponse"] = "BROADCAST_RESPONSE";
})(TempleDAppMessageType = exports.TempleDAppMessageType || (exports.TempleDAppMessageType = {}));
var TempleDAppErrorType;
(function (TempleDAppErrorType) {
    TempleDAppErrorType["NotGranted"] = "NOT_GRANTED";
    TempleDAppErrorType["NotFound"] = "NOT_FOUND";
    TempleDAppErrorType["InvalidParams"] = "INVALID_PARAMS";
    TempleDAppErrorType["TezosOperation"] = "TEZOS_OPERATION";
})(TempleDAppErrorType = exports.TempleDAppErrorType || (exports.TempleDAppErrorType = {}));
var TemplePageMessageType;
(function (TemplePageMessageType) {
    TemplePageMessageType["Request"] = "TEMPLE_PAGE_REQUEST";
    TemplePageMessageType["Response"] = "TEMPLE_PAGE_RESPONSE";
    TemplePageMessageType["ErrorResponse"] = "TEMPLE_PAGE_ERROR_RESPONSE";
})(TemplePageMessageType = exports.TemplePageMessageType || (exports.TemplePageMessageType = {}));
