// crypto-exchange-rate/dist/model/coinmarketcap/BatchExchangeRateService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.batchExchangeRateService = void 0;
const axios_1 = __importDefault(require("axios"));
function batchExchangeRateService(apiUri, apiKey) {
    const client = axios_1.default.create({
        baseURL: apiUri,
        headers: {
            'X-CMC_PRO_API_KEY': apiKey,
            Accept: 'application/json',
            'Accept-Encoding': 'deflate, gzip',
        },
    });
    return {
        async getBatchedRate(currencies) {
            const res = await client.get('/v1/cryptocurrency/quotes/latest', {
                params: {
                    id: currencies.join(','),
                },
            });
            const data = res.data;
            if (data.status.error_code != 0) {
                throw Error(`failed to fetch exchange rate ${data.status.error_message}`);
            }
            return data;
        },
    };
}
exports.batchExchangeRateService = batchExchangeRateService;
