// crypto-exchange-rate/dist/model/coinmarketcap/CMCId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CMCId = void 0;
var CMCId;
(function (CMCId) {
    CMCId[CMCId["XPNET"] = 11752] = "XPNET";
    CMCId[CMCId["EGLD"] = 6892] = "EGLD";
    CMCId[CMCId["HT"] = 2502] = "HT";
    CMCId[CMCId["BNB"] = 1839] = "BNB";
    CMCId[CMCId["ETH"] = 1027] = "ETH";
    CMCId[CMCId["AVAX"] = 5805] = "AVAX";
    CMCId[CMCId["MATIC"] = 3890] = "MATIC";
    CMCId[CMCId["FTM"] = 3513] = "FTM";
    CMCId[CMCId["ONE"] = 3945] = "ONE";
    CMCId[CMCId["TRX"] = 1958] = "TRX";
    CMCId[CMCId["CELO"] = 5567] = "CELO";
    CMCId[CMCId["ONT"] = 2566] = "ONT";
    CMCId[CMCId["XDAI"] = 8635] = "XDAI";
    CMCId[CMCId["ALGO"] = 4030] = "ALGO";
    CMCId[CMCId["FUSE"] = 5634] = "FUSE";
    CMCId[CMCId["XTZ"] = 2011] = "XTZ";
    CMCId[CMCId["VLX"] = 4747] = "VLX";
    CMCId[CMCId["IOTX"] = 2777] = "IOTX";
    CMCId[CMCId["AURORA"] = 1027] = "AURORA";
    CMCId[CMCId["VET"] = 3077] = "VET";
    CMCId[CMCId["CKB"] = 4948] = "CKB";
    CMCId[CMCId["GT"] = 4269] = "GT";
    CMCId[CMCId["SCRT"] = 5604] = "SCRT";
    //TODO: Add OPL Here
})(CMCId = exports.CMCId || (exports.CMCId = {}));
