// @vechain/connex-driver/dist/cache.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cache = void 0;
const LRU = require("lru-cache");
const bignumber_js_1 = require("bignumber.js");
const bloom_1 = require("./bloom");
const WINDOW_LEN = 12;
class Cache {
    constructor() {
        this.irreversible = {
            blocks: new LRU(256),
            txs: new LRU(512),
            receipts: new LRU(512)
        };
        this.window = [];
    }
    handleNewBlock(head, bloom, block) {
        while (this.window.length > 0) {
            const top = this.window[this.window.length - 1];
            if (top.id === head.id) {
                return;
            }
            if (top.id === head.parentID) {
                break;
            }
            this.window.pop();
        }
        this.window.push(Object.assign(Object.assign({}, head), { bloom: bloom ? bloom_1.newFilter(Buffer.from(bloom.bits.slice(2), 'hex'), bloom.k) : undefined, block, accounts: new Map(), txs: new Map(), receipts: new Map(), tied: new Map() }));
        // shift out old slots and move cached items into frozen cache
        while (this.window.length > WINDOW_LEN) {
            const bottom = this.window.shift();
            bottom.txs.forEach((v, k) => this.irreversible.txs.set(k, v));
            bottom.receipts.forEach((v, k) => this.irreversible.receipts.set(k, v));
            if (bottom.block) {
                this.irreversible.blocks.set(bottom.block.id, bottom.block);
                this.irreversible.blocks.set(bottom.block.number, bottom.block);
            }
        }
    }
    getBlock(revision, fetch) {
        return __awaiter(this, void 0, void 0, function* () {
            let block = this.irreversible.blocks.get(revision) || null;
            if (block) {
                return block;
            }
            const { slot } = this.findSlot(revision);
            if (slot && slot.block) {
                return slot.block;
            }
            block = yield fetch();
            if (block) {
                if (slot && slot.id === block.id) {
                    slot.block = block;
                }
                if (this.isIrreversible(block.number)) {
                    this.irreversible.blocks.set(block.id, block);
                    if (block.isTrunk) {
                        this.irreversible.blocks.set(block.number, block);
                    }
                }
            }
            return block;
        });
    }
    getTx(txid, fetch) {
        return __awaiter(this, void 0, void 0, function* () {
            let tx = this.irreversible.txs.get(txid) || null;
            if (tx) {
                return tx;
            }
            for (const slot of this.window) {
                tx = slot.txs.get(txid) || null;
                if (tx) {
                    return tx;
                }
            }
            tx = yield fetch();
            if (tx && tx.meta) { // only cache non-pending tx
                const { slot } = this.findSlot(tx.meta.blockID);
                if (slot) {
                    slot.txs.set(txid, tx);
                }
                if (this.isIrreversible(tx.meta.blockNumber)) {
                    this.irreversible.txs.set(txid, tx);
                }
            }
            return tx;
        });
    }
    getReceipt(txid, fetch) {
        return __awaiter(this, void 0, void 0, function* () {
            let receipt = this.irreversible.receipts.get(txid) || null;
            if (receipt) {
                return receipt;
            }
            for (const slot of this.window) {
                receipt = slot.receipts.get(txid) || null;
                if (receipt) {
                    return receipt;
                }
            }
            receipt = yield fetch();
            if (receipt) {
                const { slot } = this.findSlot(receipt.meta.blockID);
                if (slot) {
                    slot.receipts.set(txid, receipt);
                }
                if (this.isIrreversible(receipt.meta.blockNumber)) {
                    this.irreversible.receipts.set(txid, receipt);
                }
            }
            return receipt;
        });
    }
    getAccount(addr, revision, fetch) {
        return __awaiter(this, void 0, void 0, function* () {
            const found = this.findSlot(revision);
            for (let i = found.index; i >= 0; i--) {
                const slot = this.window[i];
                const acc = slot.accounts.get(addr);
                if (acc) {
                    if (i !== found.index) {
                        found.slot.accounts.set(addr, acc);
                    }
                    return acc.snapshot(found.slot.timestamp);
                }
                if (!slot.bloom || testBytesHex(slot.bloom, addr)) {
                    // account might be dirty
                    break;
                }
            }
            const accObj = yield fetch();
            if (found.slot) {
                found.slot.accounts.set(addr, new Account(accObj, found.slot.timestamp));
            }
            return accObj;
        });
    }
    /**
     * get cached entry which is tied to a batch of addresses
     * @param key the cache key
     * @param revision block id where cache bound to
     * @param fetch to fetch value when cache missing
     * @param hints array of tied addresses, as the gist to invalidate cache key. undefined means the key is always
     * invalidated on different revision.
     */
    getTied(key, revision, fetch, hints) {
        return __awaiter(this, void 0, void 0, function* () {
            const found = this.findSlot(revision);
            for (let i = found.index; i >= 0; i--) {
                const slot = this.window[i];
                const v = slot.tied.get(key);
                if (v) {
                    if (i !== found.index) {
                        found.slot.tied.set(key, v);
                    }
                    return v;
                }
                if (!slot.bloom || !hints) {
                    break;
                }
                // if hints.length === 0, never invalidate cache
                if (hints.some(t => testBytesHex(slot.bloom, t))) {
                    // might be dirty
                    break;
                }
            }
            const value = yield fetch();
            if (found.slot) {
                found.slot.tied.set(key, value);
            }
            return value;
        });
    }
    findSlot(revision) {
        const index = this.window.findIndex(s => s.id === revision || s.number === revision);
        if (index >= 0) {
            return { slot: this.window[index], index };
        }
        return { index };
    }
    isIrreversible(n) {
        if (this.window.length > 0) {
            return n < this.window[this.window.length - 1].number - WINDOW_LEN;
        }
        return false;
    }
}
exports.Cache = Cache;
function testBytesHex(filter, hex) {
    let buf = Buffer.from(hex.slice(2), 'hex');
    const nzIndex = buf.findIndex(v => v !== 0);
    if (nzIndex < 0) {
        buf = Buffer.alloc(0);
    }
    else {
        buf = buf.slice(nzIndex);
    }
    return filter.contains(buf);
}
const ENERGY_GROWTH_RATE = 5000000000;
class Account {
    constructor(obj, initTimestamp) {
        this.obj = obj;
        this.initTimestamp = initTimestamp;
    }
    snapshot(timestamp) {
        return Object.assign(Object.assign({}, this.obj), { energy: this.energyAt(timestamp) });
    }
    energyAt(timestamp) {
        if (timestamp < this.initTimestamp) {
            return this.obj.energy;
        }
        return '0x' + new bignumber_js_1.default(this.obj.balance)
            .times(timestamp - this.initTimestamp)
            .times(ENERGY_GROWTH_RATE)
            .dividedToIntegerBy(1e18)
            .plus(this.obj.energy)
            .toString(16);
    }
}
