// @vechain/connex-driver/dist/promint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterruptedError = exports.PromInt = void 0;
/** class to make promise interruptable */
class PromInt {
    constructor() {
        this.rejectors = new Set();
    }
    /**
     * interrupt all wrapped promises
     */
    interrupt() {
        const rejectors = this.rejectors;
        this.rejectors = new Set();
        rejectors.forEach(r => r(new InterruptedError()));
    }
    /**
     * wrap the promise
     * @param p the given promise
     * @returns the wrapped promise which will raise InterruptedError on interruption
     */
    wrap(p) {
        return new Promise((resolve, reject) => {
            const rejectors = this.rejectors;
            rejectors.add(reject);
            void p.then(resolve)
                .catch(reject)
                .then(() => rejectors.delete(reject));
        });
    }
}
exports.PromInt = PromInt;
class InterruptedError extends Error {
    constructor() {
        super('promise interrupted');
    }
}
exports.InterruptedError = InterruptedError;
InterruptedError.prototype.name = 'InterruptedError';
