// @vechain/connex-driver/dist/simple-websocket-reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleWebSocketReader = void 0;
const WebSocket = require("isomorphic-ws");
class SimpleWebSocketReader {
    constructor(url, timeout = 30 * 1000) {
        this.timeout = timeout;
        this.callbacks = [];
        this.ws = new WebSocket(url);
        this.ws.onmessage = ev => {
            try {
                const cbs = this.callbacks;
                this.callbacks = [];
                cbs.forEach(cb => cb(ev.data));
            }
            catch (err) {
                this.setError(err);
                this.ws.close();
            }
        };
        this.ws.onerror = ev => {
            this.setError(ev.error);
            this.ws.close();
        };
        this.ws.onclose = () => {
            this.setError(new Error('closed'));
        };
    }
    read() {
        return new Promise((resolve, reject) => {
            if (this.error) {
                return reject(this.error);
            }
            const timer = setTimeout(() => {
                reject(new Error('ws read timeout'));
            }, this.timeout);
            this.callbacks.push((data, err) => {
                clearTimeout(timer);
                if (err) {
                    return reject(err);
                }
                resolve(data);
            });
        });
    }
    close() {
        this.ws.close();
    }
    setError(err) {
        if (!this.error) {
            this.error = err;
            const cbs = this.callbacks;
            this.callbacks = [];
            cbs.forEach(cb => cb(null, err));
        }
    }
}
exports.SimpleWebSocketReader = SimpleWebSocketReader;
