// @vechain/connex-framework/dist/explainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.newExplainer = void 0;
const revert_reason_1 = require("./revert-reason");
const R = require("./rules");
const bignumber_js_1 = require("bignumber.js");
function newExplainer(readyDriver, clauses) {
    const opts = {};
    let cacheHints;
    return {
        caller(addr) {
            opts.caller = R.test(addr, R.address, 'arg0').toLowerCase();
            return this;
        },
        gas(gas) {
            opts.gas = R.test(gas, R.uint64, 'arg0');
            return this;
        },
        gasPrice(gp) {
            opts.gasPrice = R.test(gp, R.bigInt, 'arg0').toString().toLowerCase();
            return this;
        },
        gasPayer(addr) {
            opts.gasPayer = R.test(addr, R.address, 'arg0').toLowerCase();
            return this;
        },
        cache(hints) {
            cacheHints = R.test(hints, [R.address], 'arg0').map(t => t.toLowerCase());
            return this;
        },
        execute() {
            const transformedClauses = clauses.map(c => {
                return {
                    to: c.to ? c.to.toLowerCase() : null,
                    value: new bignumber_js_1.default(c.value).toString(10),
                    data: (c.data || '0x').toLowerCase()
                };
            });
            return readyDriver.then(d => d.explain(Object.assign({ clauses: transformedClauses }, opts), d.head.id, cacheHints))
                .then(outputs => {
                return outputs.map(o => {
                    if (o.reverted) {
                        const revertReason = revert_reason_1.decodeRevertReason(o.data);
                        return Object.assign(Object.assign({}, o), { revertReason });
                    }
                    return o;
                });
            });
        }
    };
}
exports.newExplainer = newExplainer;
