// @vechain/connex-framework/dist/rules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.string = exports.address = exports.hexString = exports.bigInt = exports.bool = exports.uint64 = exports.uint32 = exports.uint8 = exports.bytes32 = exports.bytes8 = exports.bytes = exports.test = exports.ensure = exports.BadParameter = exports.isBigInt = exports.isUInt = exports.isHexBytes = exports.isHexString = exports.isDecString = void 0;
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
function isDecString(val) {
    return typeof val === 'string' && /^[0-9]+$/.test(val);
}
exports.isDecString = isDecString;
function isHexString(val) {
    return typeof val === 'string' && /^0x[0-9a-f]+$/i.test(val);
}
exports.isHexString = isHexString;
function isHexBytes(val, n) {
    if (typeof val !== 'string' || !/^0x[0-9a-f]*$/i.test(val)) {
        return false;
    }
    return n ? val.length === n * 2 + 2 : val.length % 2 === 0;
}
exports.isHexBytes = isHexBytes;
function isUInt(val, bit) {
    if (val < 0 || !Number.isInteger(val)) {
        return false;
    }
    return bit ? val < Math.pow(2, bit) : true;
}
exports.isUInt = isUInt;
function isBigInt(v) {
    return typeof v === 'string' ?
        (isDecString(v) || isHexString(v)) :
        isUInt(v, 0);
}
exports.isBigInt = isBigInt;
class BadParameter extends Error {
    constructor(msg) {
        super(msg);
    }
}
exports.BadParameter = BadParameter;
BadParameter.prototype.name = 'BadParameter';
function ensure(b, msg) {
    if (!b) {
        throw new BadParameter(msg);
    }
}
exports.ensure = ensure;
const V = require("validator-ts");
function test(value, scheme, context) {
    try {
        return V.validate(value, scheme, context);
    }
    catch (err) {
        if (err instanceof V.ValidationError) {
            throw new BadParameter(err.message);
        }
        throw err;
    }
}
exports.test = test;
// rules
function bytes(v) {
    return isHexBytes(v) ? '' : 'expected bytes in hex string';
}
exports.bytes = bytes;
function bytes8(v) {
    return isHexBytes(v, 8) ? '' : 'expected bytes8';
}
exports.bytes8 = bytes8;
function bytes32(v) {
    return isHexBytes(v, 32) ? '' : 'expected bytes32';
}
exports.bytes32 = bytes32;
function uint8(v) {
    return isUInt(v, 8) ? '' : 'expected 8-bit unsigned integer';
}
exports.uint8 = uint8;
function uint32(v) {
    return isUInt(v, 32) ? '' : 'expected 32-bit unsigned integer';
}
exports.uint32 = uint32;
function uint64(v) {
    return isUInt(v, 64) ? '' : 'expected 64-bit unsigned integer';
}
exports.uint64 = uint64;
function bool(v) {
    return typeof v === 'boolean' ? '' : 'expected boolean';
}
exports.bool = bool;
function bigInt(v) {
    return isBigInt(v) ? '' : 'expected unsigned integer in number or string';
}
exports.bigInt = bigInt;
function hexString(v) {
    return isHexString(v) ? '' : 'expected integer in hex string';
}
exports.hexString = hexString;
function address(v) {
    return isHexBytes(v, 20) ? '' : 'expected address';
}
exports.address = address;
function string(v) {
    return typeof v === 'string' ? '' : 'expected string';
}
exports.string = string;
